src/math/dd/dd_bdd.cpp

```diff
@@ -1014,14 +996,133 @@ namespace dd {
+    bddv bdd_manager::mk_add(bddv const& a, bddv const& b) {
         SASSERT(a.size() == b.size());
         bdd carry = mk_false();
-        bddv result;
+        bddv result(this);
+#if 0
         for (unsigned i = 0; i < a.size(); ++i) {
+            result.push_back(carry ^ a[i] ^ b[i]);
             carry = (carry && a[i]) || (carry && b[i]) || (a[i] && b[i]);
+        }
+#else
+        if (a.size() > 0) {
+            result.push_back(a[0] ^ b[0]);
+        }
+        for (unsigned i = 1; i < a.size(); ++i) {
+            carry = (carry && a[i-1]) || (carry && b[i-1]) || (a[i-1] && b[i-1]);
             result.push_back(carry ^ a[i] ^ b[i]);
         }
+#endif
+        return result;
+    }
+
+    void bdd_manager::bddv_shl(bddv &a) {
+        for (unsigned j = a.size(); j-- > 1; ) {
+            a[j] = a[j - 1];
+        }
+        a[0] = mk_false();
+    }
+
+    bddv bdd_manager::mk_mul(bddv const& a, bddv const& b) {
+        SASSERT(a.size() == b.size());
+        bddv a_shifted = a;
+        bddv result = mk_zero(a.size());
+        for (unsigned i = 0; i < b.size(); ++i) {
+#if 1
+            bddv s = a_shifted;
+            for (unsigned j = i; j < b.size(); ++j) {
+                s[j] &= b[i];
+            }
+            result = mk_add(result, s);
+#else
+            // From BuDDy's bvec_mul. It seems to compute more intermediate BDDs than the version above?
+            bddv added = mk_add(result, a_shifted);
+            for (unsigned j = 0; j < result.size(); ++j) {
+                result[j] = mk_ite(b[i], added[j], result[j]);
+            }
+#endif
+            bddv_shl(a_shifted);
+        }
+        return result;
+    }
+
+    template <class GetBitFn>
+    bddv bdd_manager::mk_mul(bddv const& a, GetBitFn get_bit) {
+        bddv a_shifted = a;
+        bddv result = mk_zero(a.size());
+        for (unsigned i = 0; i < a.size(); ++i) {
+            if (get_bit(i)) {
+                result = mk_add(result, a_shifted);
+            }
+            bddv_shl(a_shifted);
+        }
+        return result;
+    }
+
+    bddv bdd_manager::mk_mul(bddv const& a, rational const& val) {
+        SASSERT(val.is_int() && val >= 0 && val < rational::power_of_two(a.size()));
+        return mk_mul(a, [val](unsigned i) { return val.get_bit(i); });
+    }
+
+    bddv bdd_manager::mk_mul(bddv const& a, bool_vector const& b) {
+        SASSERT(a.size() == b.size());
+        return mk_mul(a, [b](unsigned i) { return b[i]; });
+    }
+
+    bddv bdd_manager::mk_num(rational const& n, unsigned num_bits) {
+        SASSERT(n.is_int() && n >= 0 && n < rational::power_of_two(num_bits));
+        bddv result(this);
+        for (unsigned i = 0; i < num_bits; ++i) {
+            result.push_back(n.get_bit(i) ? mk_true() : mk_false());
+        }
+        return result;
+    }
+
+    bddv bdd_manager::mk_ones(unsigned num_bits) {
+        bddv result(this);
+        for (unsigned i = 0; i < num_bits; ++i) {
```

Comment:
after years of instisting on { for single statements I have abandoned this style. New code can avoid extra { }.
The benefits are too rare: accidentially adding an extra statement to be grouped inside of the for loop or having a poorly defined macro that has more than one statement.

Reply:
Ok. I'm trying to follow your style but often the extra { } sneak in out of habit :)